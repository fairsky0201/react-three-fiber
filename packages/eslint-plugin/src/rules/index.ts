// THIS FILE WAS GENERATED DO NOT MODIFY BY HAND
// @command yarn codegen:eslint

import noCloneInFrameLoop from './no-clone-in-frame-loop'

export default {
  'no-clone-in-frame-loop': noCloneInFrameLoop,
}
