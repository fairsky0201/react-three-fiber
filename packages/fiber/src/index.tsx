// required for react-test-renderer
export { createRenderer } from './core/renderer'
export { createStore } from './core/store'

export * from './web'
