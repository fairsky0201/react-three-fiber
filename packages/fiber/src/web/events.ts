import * as THREE from 'three'
import { UseStore } from 'zustand'
import { RootState } from '../core/store'
import type { DomEvent, EventManager, Intersection, ThreeEvent } from '../core/events'
import { createEvents, EventHandlers } from '../core/events'
import { Instance } from '../core/renderer'

export function createDOMEvents(store: UseStore<RootState>): EventManager<HTMLElement> {
  const temp = new THREE.Vector3()
  const { hovered, makeId, prepareRay, intersect, patchIntersects } = createEvents(store)

  function calculateDistance(event: DomEvent) {
    const { internal } = store.getState()
    const dx = event.offsetX - internal.initialClick[0]
    const dy = event.offsetY - internal.initialClick[1]
    return Math.round(Math.sqrt(dx * dx + dy * dy))
  }

  /**  Handles intersections by forwarding them to handlers */
  function handleIntersects(intersections: Intersection[], event: DomEvent, callback: (event: DomEvent) => void) {
    const { raycaster, mouse, camera, internal } = store.getState()
    // If anything has been found, forward it to the event listeners
    if (intersections.length) {
      const unprojectedPoint = temp.set(mouse.x, mouse.y, 0).unproject(camera)
      const delta = event.type === 'click' ? calculateDistance(event) : 0
      const releasePointerCapture = (id: number) => (event.target as Element).releasePointerCapture(id)
      const localState = { stopped: false, captured: false }

      for (const hit of intersections) {
        const setPointerCapture = (id: number) => {
          // If the hit is going to be captured flag that we're in captured state
          if (!localState.captured) {
            localState.captured = true
            // The captured hit array is reset to collect hits
            internal.captured = []
          }
          // Push hits to the array
          if (internal.captured)
            internal.captured.push(hit)
            // Call the original event now
          ;(event.target as Element).setPointerCapture(id)
        }

        let raycastEvent: any = {
          ...hit,
          intersections,
          stopped: localState.stopped,
          delta,
          unprojectedPoint,
          ray: raycaster.ray,
          camera: camera,
          // Hijack stopPropagation, which just sets a flag
          stopPropagation: () => {
            // https://github.com/pmndrs/react-three-fiber/issues/596
            // Events are not allowed to stop propagation if the pointer has been captured
            const cap = internal.captured
            if (!cap || cap.find((h) => h.eventObject.id === hit.eventObject.id)) {
              raycastEvent.stopped = localState.stopped = true

              // Propagation is stopped, remove all other hover records
              // An event handler is only allowed to flush other handlers if it is hovered itself
              if (hovered.size && Array.from(hovered.values()).find((i) => i.eventObject === hit.eventObject)) {
                // Objects cannot flush out higher up objects that have already caught the event
                const higher = intersections.slice(0, intersections.indexOf(hit))
                handlePointerCancel(raycastEvent as DomEvent, [...higher, hit])
              }
            }
          },
          target: { ...event.target, setPointerCapture, releasePointerCapture },
          currentTarget: { ...event.currentTarget, setPointerCapture, releasePointerCapture },
          sourceEvent: event,
        }

        // Add native event props
        for (let prop of Object.keys(Object.getPrototypeOf(event))) raycastEvent[prop] = event[prop as keyof DomEvent]
        // Call subscribers
        callback(raycastEvent as DomEvent)
        // Event bubbling may be interrupted by stopPropagation
        if (localState.stopped === true) break
      }
    }
    return intersections
  }

  function handlePointerCancel(event: DomEvent, hits?: Intersection[], prepare = true) {
    if (prepare) prepareRay(event)

    Array.from(hovered.values()).forEach((hoveredObj) => {
      // When no objects were hit or the the hovered object wasn't found underneath the cursor
      // we call onPointerOut and delete the object from the hovered-elements map
      if (
        hits &&
        (!hits.length || !hits.find((hit) => hit.object === hoveredObj.object && hit.index === hoveredObj.index))
      ) {
        const eventObject = hoveredObj.eventObject
        const handlers = ((eventObject as unknown) as Instance).__r3f.handlers
        hovered.delete(makeId(hoveredObj))
        if (handlers) {
          // Clear out intersects, they are outdated by now
          const data = { ...hoveredObj, intersections: hits || [] }
          handlers.onPointerOut?.(data as ThreeEvent<PointerEvent>)
          handlers.onPointerLeave?.(data as ThreeEvent<PointerEvent>)
        }
      }
    })
  }

  function pointerEventsOnly(objects: THREE.Object3D[]) {
    return objects.filter((obj) =>
      ['Move', 'Over', 'Enter', 'Out', 'Leave'].some(
        (name) => ((obj as unknown) as Instance).__r3f.handlers?.[('onPointer' + name) as keyof EventHandlers],
      ),
    )
  }

  function handlePointerMove(event: DomEvent, prepare = true) {
    if (prepare) prepareRay(event)
    // Get fresh intersects
    const hits = patchIntersects(intersect(pointerEventsOnly), event)
    // Take care of unhover
    handlePointerCancel(event, hits)
    handleIntersects(hits, event, (data: DomEvent) => {
      const eventObject = data.eventObject
      const handlers = ((eventObject as unknown) as Instance).__r3f.handlers
      // Check presence of handlers
      if (!handlers) return
      // Check if mouse enter or out is present
      if (handlers.onPointerOver || handlers.onPointerEnter || handlers.onPointerOut || handlers.onPointerLeave) {
        const id = makeId(data)
        const hoveredItem = hovered.get(id)
        if (!hoveredItem) {
          // If the object wasn't previously hovered, book it and call its handler
          hovered.set(id, data)
          handlers.onPointerOver?.(data as ThreeEvent<PointerEvent>)
          handlers.onPointerEnter?.(data as ThreeEvent<PointerEvent>)
        } else if (hoveredItem.stopped) {
          // If the object was previously hovered and stopped, we shouldn't allow other items to proceed
          data.stopPropagation()
        }
      }
      // Call mouse move
      handlers.onPointerMove?.(data as ThreeEvent<PointerEvent>)
    })
    return hits
  }

  function pointerMissed(event: MouseEvent, objects: THREE.Object3D[]) {
    objects.forEach((object: THREE.Object3D) =>
      ((object as unknown) as Instance).__r3f.handlers?.onPointerMissed?.(event as ThreeEvent<PointerEvent>),
    )
  }

  const handlePointer = (name: string) => (event: DomEvent, prepare = true) => {
    const { onPointerMissed, internal } = store.getState()

    if (prepare) prepareRay(event)
    // Get fresh intersects
    const hits = patchIntersects(intersect(), event)
    handleIntersects(hits, event, (data: DomEvent) => {
      const eventObject = data.eventObject
      const handlers = ((eventObject as unknown) as Instance).__r3f.handlers
      const handler = handlers?.[name as keyof EventHandlers] as (event: ThreeEvent<PointerEvent>) => void
      if (handler) {
        // Forward all events back to their respective handlers with the exception of click events,
        // which must use the initial target
        if (
          (name !== 'onClick' && name !== 'onContextMenu' && name !== 'onDoubleClick') ||
          internal.initialHits.includes(eventObject)
        ) {
          handler(data as ThreeEvent<PointerEvent>)
          pointerMissed(
            event,
            internal.interaction.filter((object) => object !== eventObject),
          )
        }
      }
    })
    // If a click yields no results, pass it back to the user as a miss
    if (name === 'onPointerDown') {
      internal.initialClick = [event.offsetX, event.offsetY]
      internal.initialHits = hits.map((hit) => hit.eventObject)
    }

    if ((name === 'onClick' || name === 'onContextMenu' || name === 'onDoubleClick') && !hits.length) {
      if (calculateDistance(event) <= 2) {
        pointerMissed(event, internal.interaction)
        if (onPointerMissed) onPointerMissed()
      }
    }
  }

  const mapNames = {
    onClick: 'click',
    onContextMenu: 'contextmenu',
    onDoubleClick: 'dblclick',
    onWheel: 'wheel',
    onPointerDown: 'pointerdown',
    onPointerUp: 'pointerup',
    onPointerLeave: 'pointerleave',
    onPointerMove: 'pointermove',
    onPointerCancel: 'pointercancel',
    onLostPointerCapture: 'lostpointercapture',
  }

  return {
    connected: false,
    handlers: {
      onClick: handlePointer('onClick') as EventListener,
      onContextMenu: handlePointer('onContextMenu') as EventListener,
      onDoubleClick: handlePointer('onDoubleClick') as EventListener,
      onWheel: handlePointer('onWheel') as EventListener,
      onPointerDown: handlePointer('onPointerDown') as EventListener,
      onPointerUp: handlePointer('onPointerUp') as EventListener,
      onPointerLeave: ((e: any) => handlePointerCancel(e, [])) as EventListener,
      onPointerCancel: ((e: any) => handlePointerCancel(e, [])) as EventListener,
      onPointerMove: (handlePointerMove as unknown) as EventListener,
      onLostPointerCapture: ((e: any) => (
        (store.getState().internal.captured = undefined), handlePointerCancel(e)
      )) as EventListener,
    },
    connect: <TElement extends HTMLElement>(target: TElement) => {
      const { set, events } = store.getState()
      events.disconnect?.()
      set((state) => ({ events: { ...state.events, connected: target } }))
      Object.entries(events?.handlers ?? []).forEach(([name, event]) =>
        target.addEventListener(mapNames[name as keyof typeof mapNames], event, { passive: true }),
      )
    },
    disconnect: () => {
      const { set, events } = store.getState()
      if (events.connected) {
        Object.entries(events.handlers ?? []).forEach(([name, event]) => {
          if (events && events.connected instanceof HTMLElement) {
            events.connected.removeEventListener(mapNames[name as keyof typeof mapNames], event)
          }
        })
        set((state) => ({ events: { ...state.events, connected: null } }))
      }
    },
  }
}
