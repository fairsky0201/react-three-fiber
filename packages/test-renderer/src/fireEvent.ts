import ReactReconciler from 'react-reconciler'
import type { Camera } from 'react-three-fiber'

import { toEventHandlerName } from './helpers/strings'

import type { MockSceneChild, MockUseStoreState } from './createMockStore'

/**
 * this is an empty object of any,
 * the data is passed to a new event
 * and subsequently passed to the
 * event handler you're calling
 */
export type MockEventData = {
  [key: string]: any
}

export type MockSyntheticEvent = {
  camera: Camera
  stopPropagation: () => void
  target: MockSceneChild
  currentTarget: MockSceneChild
  sourceEvent: MockEventData
  [key: string]: any
}

export const createEventFirer = (
  act: ReactReconciler.Reconciler<unknown, unknown, unknown, unknown, unknown>['act'],
  store: MockUseStoreState,
) => {
  const findEventHandler = (
    element: MockSceneChild,
    eventName: string,
  ): ((event: MockSyntheticEvent) => any) | null => {
    const eventHandlerName = toEventHandlerName(eventName)

    const props = element.__r3f.memoizedProps

    if (typeof props[eventHandlerName] === 'function') {
      return props[eventHandlerName]
    }

    if (typeof props[eventName] === 'function') {
      return props[eventName]
    }

    return null
  }

  const createSyntheticEvent = (element: MockSceneChild, data: MockEventData): MockSyntheticEvent => {
    const raycastEvent = {
      camera: store.getState().camera,
      stopPropagation: () => {},
      target: { ...element },
      currentTarget: { ...element },
      sourceEvent: data,
      ...data,
    }
    return raycastEvent
  }

  const invokeEvent = async (element: MockSceneChild, eventName: string, data: MockEventData): Promise<any> => {
    const handler = findEventHandler(element, eventName)

    if (!handler) {
      return
    }

    let returnValue: any

    await act(async () => {
      returnValue = handler(createSyntheticEvent(element, data))
    })

    return returnValue
  }

  const fireEvent = async (element: MockSceneChild, eventName: string, data: MockEventData = {}): Promise<any> =>
    await invokeEvent(element, eventName, data)

  return fireEvent
}
