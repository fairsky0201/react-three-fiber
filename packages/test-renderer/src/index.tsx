import * as THREE from 'three'
import * as React from 'react'

import { is } from 'react-three-fiber/src/core/is'
import { createRenderer, Root } from 'react-three-fiber/src/core/renderer'

import {
  createMockStore,
  MockStoreProps,
  MockUseStoreState,
  context,
  MockScene,
  MockSceneChild,
} from './createMockStore'
import { createCanvas, CreateCanvasParameters } from './createTestCanvas'

type RenderProps = Omit<MockStoreProps, 'gl' | 'context'> & {
  gl?: THREE.WebGLRenderer | THREE.WebGLRendererParameters
  concurrent?: boolean
}

type MockRoot = Omit<Root, 'store'> & {
  store: MockUseStoreState
}

interface ReactThreeTestRendererSceneGraphItem {
  type: string
  name: string
  children: ReactThreeTestRendererSceneGraphItem[] | null
}

interface ReactThreeTestRendererTreeNode {
  type: string
  props: {
    [key: string]: any
  }
  children: ReactThreeTestRendererTreeNode[]
}

export type ReactThreeTestRendererSceneGraph = ReactThreeTestRendererSceneGraphItem[]

export type ReactThreeTestRendererTree = ReactThreeTestRendererTreeNode

export interface ReactThreeTestRendereOptions extends CreateCanvasParameters {}

const mockRoots = new Map<any, MockRoot>()
const { reconciler, applyProps } = createRenderer(mockRoots)

const render = <TRootNode,>(
  element: React.ReactNode,
  id: TRootNode,
  { gl, size, concurrent, ...props }: RenderProps = { size: { width: 0, height: 0 } },
): THREE.Scene => {
  let root = mockRoots.get(id)
  let fiber = root?.fiber
  let store = root?.store
  let state = store?.getState()

  if (fiber && state) {
    const lastProps = state.internal.lastProps

    // When a root was found, see if any fundamental props must be changed or exchanged

    // Check pixelratio
    if (props.dpr !== undefined && !is.equ(lastProps.dpr, props.dpr)) state.setDpr(props.dpr)
    // Check size
    if (size !== undefined && !is.equ(lastProps.size, size)) state.setSize(size.width, size.height)

    // For some props we want to reset the entire root

    // Changes to the color-space
    const linearChanged = props.linear !== lastProps.linear
    if (linearChanged) {
      unmount(id)
      fiber = undefined
    }
  }

  if (!fiber) {
    // If no root has been found, make one
    store = createMockStore(applyProps, { gl: {}, size, ...props })

    fiber = reconciler.createContainer(store, concurrent ? 2 : 0, false, null)
    // Map it
    mockRoots.set(id, { fiber, store })
  }

  if (store && fiber) {
    reconciler.updateContainer(<Provider store={store} element={element} />, fiber, null, () => undefined)
    return store!.getState().scene as THREE.Scene
  } else {
    throw 'R3F: Error creating fiber-root!'
  }
}

function Provider({ store, element }: { store: MockUseStoreState; element: React.ReactNode }) {
  React.useEffect(() => store.getState().set((state) => ({ internal: { ...state.internal, active: true } })), [])
  return <context.Provider value={store}>{element}</context.Provider>
}

const unmount = <TRootNode,>(id: TRootNode) => {
  const root = mockRoots.get(id)
  const fiber = root?.fiber

  const dispose = (obj: any) => {
    if (obj.dispose && obj.type !== 'Scene') obj.dispose()
    for (const p in obj) {
      if (typeof p === 'object' && (p as any).dispose) (p as any).dispose()
      delete obj[p]
    }
  }

  if (fiber) {
    reconciler.updateContainer(null, fiber, null, () => {
      const state = root?.store.getState()
      if (state) {
        dispose(state.raycaster)
        dispose(state.camera)
        dispose(state)
      }
      mockRoots.delete(id)
    })
  }
}

const lowerCaseFirstLetter = (str: string) => `${str.charAt(0).toLowerCase()}${str.slice(1)}`

const treeObjectFactory = (
  type: ReactThreeTestRendererTreeNode['type'],
  props: ReactThreeTestRendererTreeNode['props'],
  children: ReactThreeTestRendererTreeNode['children'],
): ReactThreeTestRendererTreeNode => ({
  type,
  props,
  children,
})

const toTreeBranch = (obj: MockSceneChild[]): ReactThreeTestRendererTreeNode[] =>
  obj.map((child) => {
    return treeObjectFactory(
      lowerCaseFirstLetter(child.type || child.constructor.name),
      { ...child.__r3f.memoizedProps },
      toTreeBranch([...(child.children || []), ...child.__r3f.objects]),
    )
  })

const toTree = (scene: MockScene): ReactThreeTestRendererTree => ({
  type: 'scene',
  props: {},
  children: scene.children.map((obj) =>
    treeObjectFactory(
      lowerCaseFirstLetter(obj.type),
      { ...obj.__r3f.memoizedProps },
      toTreeBranch([...obj.children, ...obj.__r3f.objects]),
    ),
  ),
})

const graphObjectFactory = (
  type: ReactThreeTestRendererSceneGraphItem['type'],
  name: ReactThreeTestRendererSceneGraphItem['name'],
  children: ReactThreeTestRendererSceneGraphItem['children'],
): ReactThreeTestRendererSceneGraphItem => ({
  type,
  name,
  children,
})

const toGraph = (object: THREE.Object3D): ReactThreeTestRendererSceneGraphItem[] =>
  object.children.map((child) => graphObjectFactory(child.type, child.name || '', toGraph(child)))

reconciler.injectIntoDevTools({
  bundleType: process.env.NODE_ENV === 'production' ? 0 : 1,
  rendererPackageName: 'react-three-test-renderer',
  // @ts-expect-error it's a babel macro
  version: typeof R3F_VERSION !== 'undefined' ? R3F_VERSION : '0.0.0',
})

const create = (element: React.ReactNode, options?: ReactThreeTestRendereOptions) => {
  const canvas = createCanvas(options)
  const scene = render(element, canvas)

  return {
    scene,
    unmount,
    update(newElement: React.ReactNode) {
      const fiber = mockRoots.get(canvas)?.fiber
      if (fiber) {
        reconciler.updateContainer(newElement, fiber, null, () => null)
      } else {
        return
      }
    },
    toTree: () => {
      if (!scene) {
        return
      } else {
        return toTree(scene as MockScene)
      }
    },
    toGraph: () => {
      if (!scene) {
        return
      } else {
        return toGraph(scene as MockScene)
      }
    },
  }
}

const ReactThreeTestRenderer = {
  create,
  act: reconciler.act,
}

export default ReactThreeTestRenderer
