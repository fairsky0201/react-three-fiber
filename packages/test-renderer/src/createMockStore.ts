import { context, _createStore as createStore } from '@react-three/fiber'

export { createStore, context }
