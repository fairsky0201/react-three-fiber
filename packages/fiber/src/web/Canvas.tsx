import * as React from 'react'
import useMeasure, { Options as ResizeOptions } from 'react-use-measure'
import { render, unmountComponentAtNode, RenderProps } from './index'
import { createEvents as events } from './events'

export interface Props extends Omit<RenderProps, 'size'>, React.HTMLAttributes<HTMLDivElement> {
  children: React.ReactNode
  resize?: ResizeOptions
}

export function Canvas({ children, resize, style, className, ...props }: Props) {
  const [ref, size] = useMeasure({ scroll: true, debounce: { scroll: 50, resize: 0 }, ...resize })
  const canvas = React.useRef<HTMLCanvasElement>(null!)
  React.useLayoutEffect(() => {
    if (size.width > 0 && size.height > 0) {
      render(children, canvas.current, { ...props, size, events })
    }
  }, [size, children])
  React.useEffect(() => () => unmountComponentAtNode(canvas.current), [])
  return (
    <div
      ref={ref}
      className={className}
      style={{ position: 'relative', width: '100%', height: '100%', overflow: 'hidden', ...style }}>
      <canvas ref={canvas} style={{ display: 'block' }} />
    </div>
  )
}
