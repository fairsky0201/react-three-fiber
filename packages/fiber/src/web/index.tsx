import * as THREE from 'three'
import * as React from 'react'
import { RootTag } from 'react-reconciler'
import { UseStore } from 'zustand'

import { is } from '../core/is'
import { createStore, StoreProps, isRenderer, context, RootState, Size } from '../core/store'
import { createRenderer, extend, Instance, Root } from '../core/renderer'
import { createLoop, addEffect, addAfterEffect, addTail } from '../core/loop'
import { createDOMEvents as events } from './events'
import { Canvas } from './Canvas'
import { EventManager } from '../core/events'

export type RenderProps<TCanvas extends Element> = Omit<StoreProps, 'gl' | 'events' | 'size'> & {
  gl?: THREE.WebGLRenderer | THREE.WebGLRendererParameters
  events?: (store: UseStore<RootState>) => EventManager<TCanvas>
  size?: Size
  mode?: 'legacy' | 'blocking' | 'concurrent'
  onCreated?: (state: RootState) => void
}

const roots = new Map<Element, Root>()
const modes = ['legacy', 'blocking', 'concurrent']
const { invalidate, advance } = createLoop(roots)
const { reconciler, applyProps } = createRenderer(roots)

const createRendererInstance = <TElement extends Element>(
  gl: THREE.WebGLRenderer | THREE.WebGLRendererParameters | undefined,
  canvas: TElement,
): THREE.WebGLRenderer =>
  isRenderer(gl as THREE.WebGLRenderer)
    ? (gl as THREE.WebGLRenderer)
    : new THREE.WebGLRenderer({
        powerPreference: 'high-performance',
        canvas: (canvas as unknown) as HTMLCanvasElement,
        antialias: true,
        alpha: true,
        ...gl,
      })

function render<TCanvas extends Element>(
  element: React.ReactNode,
  canvas: TCanvas,
  { gl, size = { width: 0, height: 0 }, mode = 'blocking', events, onCreated, ...props }: RenderProps<TCanvas> = {},
): UseStore<RootState> {
  let root = roots.get(canvas)
  let fiber = root?.fiber
  let store = root?.store
  let state = store?.getState()

  if (fiber && state) {
    const lastProps = state.internal.lastProps

    // When a root was found, see if any fundamental props must be changed or exchanged

    // Check pixelratio
    if (props.dpr !== undefined && !is.equ(lastProps.dpr, props.dpr)) state.setDpr(props.dpr)
    // Check size
    if (size !== undefined && !is.equ(lastProps.size, size)) state.setSize(size.width, size.height)

    // For some props we want to reset the entire root

    // Changes to the color-space
    const linearChanged = props.linear !== lastProps.linear
    if (linearChanged) {
      unmountComponentAtNode(canvas)
      fiber = undefined
    }
  }

  if (!fiber) {
    // If no root has been found, make one
    // Create store
    store = createStore(applyProps, invalidate, advance, { gl: createRendererInstance(gl, canvas), size, ...props })
    const state = store.getState()
    const get = state.get
    // Create renderer
    fiber = reconciler.createContainer(store, modes.indexOf(mode) as RootTag, false, null)
    // Map it
    roots.set(canvas, { fiber, store })
    // Create and register events

    // Store events internally
    if (events) state.set({ events: events(store) })

    // VR
    if (props.vr && (gl as THREE.WebGLRenderer).xr && (gl as THREE.WebGLRenderer).setAnimationLoop) {
      ;(gl as THREE.WebGLRenderer).xr.enabled = true
      ;(gl as THREE.WebGLRenderer).setAnimationLoop((t: number) => advance(t, true, state))
    }
  }

  if (store && fiber) {
    reconciler.updateContainer(
      <Provider store={store} element={element} onCreated={onCreated} target={canvas} />,
      fiber,
      null,
      () => undefined,
    )
    return store
  } else {
    throw 'R3F: Error creating fiber-root!'
  }
}

function Provider<TElement extends Element>({
  store,
  element,
  onCreated,
  target,
}: {
  onCreated?: (state: RootState) => void
  store: UseStore<RootState>
  element: React.ReactNode
  target: TElement
}) {
  React.useEffect(() => {
    const state = store.getState()
    // Flag the canvas active, rendering will now begin
    state.set((state) => ({ internal: { ...state.internal, active: true } }))
    // Connect events
    state.events.connect?.(target)
    // Notifiy that init is completed, the scene graph exists, but nothing has yet rendered
    if (onCreated) onCreated(state)
  }, [])
  return <context.Provider value={store}>{element}</context.Provider>
}

function unmountComponentAtNode<TElement extends Element>(canvas: TElement, callback?: (canvas: TElement) => void) {
  const root = roots.get(canvas)
  const fiber = root?.fiber
  if (fiber) {
    reconciler.updateContainer(null, fiber, null, () => {
      const state = root?.store.getState()
      if (state) {
        state.events.disconnect?.()
        state.gl?.renderLists?.dispose()
        state.gl?.forceContextLoss()
        dispose((state.gl as unknown) as Instance)
        dispose((state.raycaster as unknown) as Instance)
        dispose((state.camera as unknown) as Instance)
        dispose((state as unknown) as Instance)
      }
      roots.delete(canvas)
      if (callback) callback(canvas)
    })
  }
}

function dispose(obj: Instance) {
  if (obj.dispose && obj.type !== 'Scene') obj.dispose()
  for (const p in obj) {
    if (typeof p === 'object' && (p as Instance).dispose) (p as Instance).dispose()
    delete obj[p]
  }
}

const act = reconciler.act
const hasSymbol = is.fun(Symbol) && Symbol.for
const REACT_PORTAL_TYPE = hasSymbol ? Symbol.for('react.portal') : 0xeaca
function createPortal(children: React.ReactNode, container: any, impl?: any, key: any = null): React.ReactNode {
  if (!container.__objects) container.__objects = []
  return { $$typeof: REACT_PORTAL_TYPE, key: key == null ? null : '' + key, children, container, impl }
}

reconciler.injectIntoDevTools({
  bundleType: process.env.NODE_ENV === 'production' ? 0 : 1,
  rendererPackageName: 'react-three-fiber',
  // @ts-ignore
  version: typeof R3F_VERSION !== 'undefined' ? R3F_VERSION : '0.0.0',
})

export * from '../core/hooks'
export {
  context,
  render,
  unmountComponentAtNode,
  createPortal,
  events,
  reconciler,
  applyProps,
  invalidate,
  advance,
  extend,
  addEffect,
  addAfterEffect,
  addTail,
  Canvas,
  act,
}
