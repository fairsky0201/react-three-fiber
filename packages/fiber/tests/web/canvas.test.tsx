jest.mock('scheduler', () => require('scheduler/unstable_mock'))

import * as React from 'react'
import { ReactTestRenderer, create, act } from 'react-test-renderer'
import { createCanvas } from 'react-three-test-renderer/src/createTestCanvas'
import { createWebGLContext } from 'react-three-test-renderer/src/createWebGLContext'

import { Canvas, testutil_act as r3fAct } from '../../src/web'

describe('web Canvas', () => {
  const canvas = createCanvas({
    beforeReturn: (canvas) => {
      //@ts-ignore
      canvas.getContext = (type: string) => {
        if (type === 'webgl' || type === 'webgl2') {
          return createWebGLContext(canvas)
        }
      }
    },
  })

  const createNodeMock = () => {
    return canvas
  }

  it('should correctly mount', async () => {
    let renderer: ReactTestRenderer | null = null

    await r3fAct(async () => {
      await act(async () => {
        renderer = create(
          <Canvas>
            <group />
          </Canvas>,
          {
            createNodeMock,
          },
        )
      })
    })

    const tree = renderer!.toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('should correctly unmount', async () => {
    let renderer: ReactTestRenderer | null = null

    await r3fAct(async () => {
      await act(async () => {
        renderer = create(
          <Canvas>
            <group />
          </Canvas>,
          {
            createNodeMock,
          },
        )
      })
    })

    expect(() => renderer!.unmount()).not.toThrow()
  })
})
