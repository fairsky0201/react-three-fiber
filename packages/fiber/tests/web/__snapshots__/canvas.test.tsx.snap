// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`web Canvas should correctly mount 1`] = `
<div
  style={
    Object {
      "height": "100%",
      "overflow": "hidden",
      "position": "relative",
      "width": "100%",
    }
  }
>
  <canvas
    style={
      Object {
        "display": "block",
      }
    }
  />
</div>
`;
